sycl/doc/design/OptionalDeviceFeatures.md

```diff
@@ -913,211 +913,253 @@ aspects in the array.
+offload bundle has type "spir64").
+
+However, if the user AOT compiles for an Intel GPU with
+`-fsycl-targets=spir64_gen`, the device compiler is invoked with
+`-triple spir64_gen-unknown-unknown`, which is a different triple from the case
+above.  This IR is also converted to SPIR-V and then sent to the `ocloc` tool.
+The `ocloc` tool compiles the SPIR-V to native code but also embeds the SPIR-V
+into the ocloc output.  This SPIR-V is used by the device driver if the
+application is run on a GPU device that does not match any of the AOT targets.
+
+In the new design, `ocloc` does not embed SPIR-V in its output because we
+invoke it with `-exclude_ir`.  Instead, DPC++ is responsible for embedding the
+SPIR-V in the host executable, but should that SPIR-V be generated with the
+triple `spir64-unknown-unknown` or the triple `spir64_gen-unknown-unknown`?
+We could generate both, but that seems wasteful.  If we generate the SPIR-V
+with `spir64-unknown-unknown`, does that mean we need to run the device
+compiler twice, once with `spir64_gen-unknown-unknown` (to generate SPIR-V that
+is fed to `ocloc`) and once with `spir64-unknown-unknown` (to generate SPIR-V
+that is embedded in the host executable)?  This also seems wasteful because
+it requires running the device compiler twice whenever the user AOT compiles
+for an Intel GPU.
+
+An ideal solution would be to always generate SPIR-V using the triple
+`spir64-unknown-unknown`, and use this SPIR-V as input to both `ocloc` and
+`opencl-aot`.  This would allow us to run the device compiler just once even if
+the user AOT compiles for both Intel GPU and CPU.  However, this means that the
+SPIR-V is generated using a different triple than we use today.  It's unclear
+what ramification this has.
+
+
 ### Changes to the DPC++ runtime
 
-The DPC++ runtime must be changed to check if a kernel uses any optional
-features that the device does not support.  If this happens, the runtime must
-raise a synchronous `errc::kernel_not_supported` exception.
+#### Identifying a compatible offload bundle
+
+Since the user may AOT compile for several different offload targets, the
+runtime must determine which offload bundles are compatible with the queue's
+device.  If the device is an Intel GPU, the runtime uses either
+[`ze_device_ip_version_ext_t`][9] (on Level Zero) or
+[`CL_DEVICE_IP_VERSION_INTEL`][10] (on OpenCL) to get the GMDID of the
+device.  This GMDID value can be mapped to any of the Intel GPU target names
+that are used for AOT compiled offload bundles.
+
+If the runtime doesn't find an AOT compiled offload bundle, it searches for a
+SPIR-V bundle using the offload target name "spir64".
+
+[9]: <https://spec.oneapi.io/level-zero/latest/core/api.html#ze-device-ip-version-ext-t>
+[10]: <https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_device_attribute_query.html>
+
+#### Raising an exception when features are incompatible
```

Comment:
@gmlueck could we also extend `SYCL/device requirements` in scope of this PR with the additional property set reflecting the value(-s) of `-fsycl-targets` flag?

What I mean - previously we had an [update](https://github.com/KhronosGroup/SYCL-Docs/pull/381) to clarify `is_compatible()`

> to tell if a kernel was compiled in a way that restricts the devices on which it can run

At the moment in scope of `is_compatible` we only check that the device has the same backend as the target backend of the binary image. We can't yet handle the case when -fsycl-targets=nvidia_gpu_sm_86 and the device is sm_90. To check this the image should contain the info form `-fsycl-targets`, so we probably need to put it into `device requirements`.

Reply:
Tagging @AerialMantis here because it would be good to get his perspective on how this should work with CUDA.

I was thinking that the solution I outline above for Intel GPU targets would work also for CUDA targets.  These are the relevant paragraphs from earlier in the document:

> The driver also invokes `clang-offload-wrapper` to wrap the native offload modules for inclusion in the host executable. Currently, the driver passes `-target=spir64` when not doing AOT compilation (i.e. when generating SPIR-V for the target code), and it passes `-target=spir64_gen` when doing AOT compilation. The target name `spir64_gen` indicates that the offload bundle is native code that can run on any Intel GPU.
>
> With the new design, the driver still uses `-target=spir64` to identify a SPIR-V offload bundle, however it must use a more specific device name for native offload bundles. We use the same target names accepted by `-fsycl-targets`. For example, if DPC++ is invoked with `-fsycl-targets=intel_gpu_pvc,intel_gpu_dg1`, then the driver invokes `clang-offload-wrapper` twice, once with `-target=intel_gpu_pvc` and once with `-target=intel_gpu_dg1`.

If we follow this for CUDA, it would work like this.  If the user compiles with `-fsycl-targets=nvidia_gpu_sm_86`, the compiler driver would end up bundling the offload image via `clang-offload-wrapper -target=nvidia_gpu_sm_86 ...`.  When the SYCL runtime looks for a compatible device image, it would then decide if the image's target type "nvidia_gpu_sm_86" is compatible with the queue's device.

Reply:
To clarify - do I understand it correctly, that the idea is to use not just `__SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_X86_64` as target but something like `__SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_X86_64_##DEVICE_NAME`?

Reply:
Sorry for the delay.  I wanted to talk with @mdtoguchi first to make sure I was on the right track.

The design in this document is written using the old clang offloading design (i.e. the one using `clang-offload-wrapper`), which is how DPC++ is currently implemented.  In that design, the `clang-offload-wrapper` tool accepts a command line option `-target=`, which specifies a string name for the offload target device.  We currently pass this option as "-target=spir64_gen" when compiling for an Intel GPU device.  I think this string is stored in the host ELF file in the `pi_device_binary_struct::DeviceTargetSpec` field, but I'm not certain of this.  (If you know this code better, maybe you can answer.)

My proposal is that we pass a more specific target string here, for example `-target=intel_gpu_pvc` (or maybe just `-target=pvc`).  This would allow the SYCL runtime to identify a PVC device image by looking at the `pi_device_binary_struct::DeviceTargetSpec` field.

The exact spelling of these target strings can be decided later.  For example, I think we will eventually migrate away from `-fsycl-targets=` to the standard clang option `--offload-arch=`.  When we do this, we will need to decide on the names of the Intel GPU devices that we will use with `--offload-arch=`, and we may as well use the same strings for the `pi_device_binary_struct::DeviceTargetSpec` field.

BTW, I know that we are migrating away from the old clang offloading design towards the new design that uses `clang-linker-wrapper`.  I believe this new offloading design still identifies each device image with some sort of target name, so I think the strategy I'm proposing  here still works even with that new design.

Reply:
> (If you know this code better, maybe you can answer.)

Yes, that's exactly what I was talking about. `is_compatible(Dev)` checks the value of `pi_device_binary_struct::DeviceTargetSpec` and decides if it matches the backend of the device. I didn't know about that idea:

> My proposal is that we pass a more specific target string here

That should work fine for us, thanks for the clarification.